ethstats/src/main/kotlin/org/apache/tuweni/ethstats/EthStatsServer.kt

```diff
@@ -0,0 +1,157 @@
+    }
+  }
+
+  suspend fun stop() {
+    if (started.compareAndSet(true, false)) {
+      server?.closeAwait()
+    }
+  }
+
+  fun connect(serverWebSocket: ServerWebSocket) {
+    logger.debug("New connection")
+    val websocket = serverWebSocket
+    websocket.accept()
+    websocket.exceptionHandler {
+      logger.debug("Error reading message", it)
+      websocket.close()
+    }
+    websocket.textMessageHandler {
+      try {
+        handleMessage(it, websocket)
+      } catch (e: IOException) {
+        logger.debug("Invalid payload", e)
+        websocket.close()
+      }
+    }
+  }
+
+  private fun handleMessage(message: String, websocket: ServerWebSocket) {
+    logger.debug("Received {}", message)
+    val event = mapper.readTree(message).get("emit") as ArrayNode
+    val command = event.get(0).textValue()
+    when (command) {
+      ("hello") -> {
+        val clientSecret = event.get(1).get("secret").textValue()
+        if (clientSecret != secret) {
+          logger.info("Client {} connected with wrong secret {}, disconnecting", websocket.remoteAddress(), clientSecret)
+          websocket.close()
+          return
+        }
+        websocket.writeTextMessage("{\"emit\":[\"ready\"]}")
+        val id = event.get(1).get("id").textValue()
+        val nodeInfo = mapper.readerFor(NodeInfo::class.java).readValue<NodeInfo>(event.get(1).get("info"))
+        nodeInfoReader(id, nodeInfo)
+      }
+      ("node-ping") -> {
+        logger.debug("Received a ping {}", event.get(1))
+        val clientTime = event.get(1).get("clientTime").longValue()
+        val payload = mapOf(Pair("clientTime", clientTime), Pair("serverTime", timeSupplier().toEpochMilli()))
+
+        val response = mapper.writer().writeValueAsString(mapOf(Pair("emit", listOf("node-pong", payload))))
+        EthStatsReporter.logger.debug("Sending {} message {}", command, message)
+        websocket.writeTextMessage(response)
+      }
+      ("latency") -> {
+        val id = event.get(1).get("id").textValue()
+        val latency = event.get(1).get("latency").longValue()
+        latencyReader(id, websocket.remoteAddress(), latency)
+      }
+      ("end") -> {
+        val id = event.get(1).get("id").textValue()
+        disconnectionReader(id)
+      }
+      ("stats") -> {
+        val id = event.get(1).get("id").textValue()
+        val nodeStats = mapper.readerFor(NodeStats::class.java).readValue<NodeStats>(event.get(1).get("stats"))
+        nodeStatsReader(id, nodeStats)
+      }
+      ("pending") -> {
+        val id = event.get(1).get("id").textValue()
+        val pendingTx = event.get(1).get("stats").get("pending").longValue()
+        pendingTxReader(id, pendingTx)
+      }
+      ("block") -> {
+        val id = event.get(1).get("id").textValue()
+        val block = mapper.readerFor(BlockStats::class.java).readValue<BlockStats>(event.get(1).get("block"))
+        blockReader(id, block)
+      }
+      ("update") -> {
+        // TODO implement
```

Comment:
make sense to maybe  throw a UnsupportedOperationException?

Reply:
if we throw an exception, it will bork the connection. We can add a handler for this behavior in a bit.

For now just getting stats is great.